PettingZoo wrapper

Rohan138 opened the pull request:
Hi, I'd like to contribute a [PettingZoo](https://www.pettingzoo.ml/) wrapper and a corresponding [Stable-Baselines3](https://github.com/DLR-RM/stable-baselines3) training script using PPO with parameter sharing for Meltingpot to complement the current rllib example.

We've tested the harvest_commons_open and clean_up environments with the wrapper, and it should work similarly for all environments. I can also add documentation where neccessary.

duenez commented:
Thank you for sending this pull request.

We will take a closer look later. Just a few comments from the top of my head: It seems like this change would require adding several new dependencies. You probably should make an explicit split between an RLLib example, and this PettingZoo example in setup.py. You should also update the documentation on how to run this, to improve it usability.

We'll get back to you with a detailed review in the next few days.

Thanks again!

Rohan138 commented:
Added a new setup.py extra and some documentation in the README.md

duenez commented on README.md:
```diff
@@ -112,17 +112,16 @@ NOTE: If you get a `ModuleNotFoundError: No module named 'meltingpot.python'`
+#### RLlib
+This example uses [RLLib](https://github.com/ray-project/ray) to train agents in self-play on a Melting Pot substrate.
 
 First you will need to install the dependencies needed by the examples:
```
"... needed by the RLLib example:"

duenez commented on examples/pettingzoo/meltingpot_env.py:
```diff
@@ -0,0 +1,149 @@
+import numpy as np
+from functools import lru_cache
+
+from gym.utils import EzPickle
```
Please import modules instead of objects directly, i.e.

```
from gym import utils as gym_utils
```

and later use as `gym_utils.EzPickle`

duenez commented on examples/pettingzoo/meltingpot_env.py:
```diff
@@ -0,0 +1,149 @@
+from gym.utils import EzPickle
+from pettingzoo.utils import wrappers
+from pettingzoo.utils.conversions import parallel_to_aec_wrapper
+from pettingzoo.utils.env import ParallelEnv
```
As above.

duenez commented on examples/pettingzoo/meltingpot_env.py:
```diff
@@ -0,0 +1,149 @@
+  gym_observations = {}
+  for index, observation in enumerate(timestep.observation):
+    gym_observations[PLAYER_STR_FORMAT.format(index=index)] = {
+        key: value for key, value in observation.items() if 'WORLD' not in key
```
Suggest checking if the key _starts_ with 'WORLD.', but if you want to keep it as a `not in` test, maybe at least add the dot? (i.e. 'WORLD.')

Rohan138 replied:
Yes, I think a lot of duplicated functionality between the RLlib and PettingZoo examples can be moved to utils; however, I would prefer keeping them as is here, and moving them, and adding docstrings and tests in a future PR, if that's okay.

duenez commented on examples/pettingzoo/meltingpot_env.py:
```diff
@@ -0,0 +1,149 @@
+def _remove_world_observations_from_space(
+    observation: spaces.Dict) -> spaces.Dict:
+  return spaces.Dict({
+      key: observation[key] for key in observation if 'WORLD' not in key})
```
Since this string literal is used at least twice, please move it to a module level constant, e.g.:

```
WORLD_PREFIX = 'WORLD.'
```

duenez commented on examples/pettingzoo/meltingpot_env.py:
```diff
@@ -0,0 +1,149 @@
+        return None
+    return rgb_arr
+
+class _parallel_env(MeltingPotPettingZooEnv, EzPickle):
```
Class names should be CamelCase starting with capitals. If the class is private to the module, it's OK to prefix with an underscore.

duenez commented on examples/pettingzoo/meltingpot_env.py:
```diff
@@ -0,0 +1,149 @@
+import matplotlib.pyplot as plt
+
+PLAYER_STR_FORMAT = 'player_{index}'
+MAX_CYCLES = 1000
```
It would be useful to have a comment explaining what this constant is. It seems to be the maximum number of steps of the environment? Meltingpot substrates sometimes have longer time horizons than 1000 steps. Why does it need to have a hard cutoff?

Rohan138 replied:
Appendix B in the Meltingpot paper on Substrate details states that "Episodes last 1000 steps" unless otherwise mentioned, so I replicated it for the default value.

It seems that Allelopathic Harvest is the only exceptio to this with 2000 steps, which can easily be passed in through the `max_cycles` argument to the constructor (or even `inf` for no cutoff)

duenez replied:
OK, just be aware that on future releases, this is likely to change. :)

duenez commented on examples/pettingzoo/meltingpot_env.py:
```diff
@@ -0,0 +1,149 @@
+
+  def __init__(self, env_config, max_cycles):
+    self.env_config = config_dict.ConfigDict(env_config)
+    self.max_cycles = max_cycles
```
Does this need to be public? Is it part of the PettingZoo API?

Rohan138 replied:
No, the PettingZoo public API mainly consists of the three methods `parallel_env`, `env`, and `raw_env`.

The ParallelEnv API (returned by `parallel_env`) steps all agents synchronously and is used in the example.

The `raw_env` returns an AEC environment ([Docs](https://www.pettingzoo.ml/api), [Paper](https://arxiv.org/abs/2009.14471)), which steps each agent individually.

The `env` function is `raw_env` with some thin wrappers to check that input actions are in the action space, etc.

I can make `MeltingPotPettingZooEnv` private if you'd prefer; PettingZoo recommends that you use the three functions instead of the raw environment class to initialize the environment. The _ParallelEnv class is a convenience to make `EzPickle` work.

duenez replied:
Thanks for the context. It seems then that this module really should only have 3 public members: `env`, `raw_env` and `parallel_env`. Please make everything else private. And likewise, make the all the members of classes (that make sense to do so) private.

duenez commented on examples/pettingzoo/sb3_train.py:
```diff
@@ -0,0 +1,166 @@
+from stable_baselines3.common.callbacks import EvalCallback
+from torch import nn
+
+from examples.pettingzoo.meltingpot_env import parallel_env
```
As above, import module, not functions from module directly.

duenez commented on examples/pettingzoo/meltingpot_env.py:
```diff
@@ -0,0 +1,149 @@
+  else:
+    raise ValueError('Unexpected spec: {}'.format(spec))
+
+def parallel_env(env_config, max_cycles=MAX_CYCLES):
```
This seems to be the only factory used in the training file. I suggest making only this one public, and calling it `get_parallel_env` instead. It should also have a docstring explaining what it does, and what is expected to be the properties of the returned object (also add the PyType annotation for the return).

duenez commented on examples/pettingzoo/meltingpot_env.py:
```diff
@@ -0,0 +1,149 @@
+  def reset(self):
+    """See base class."""
+    timestep = self._env.reset()
+    self.agents = self.possible_agents[:]
```
I am confused as to why the agents need to be public, and just copying from possible agents at every episode. Why couldn't it all just be one fixed tuple of agent IDs, and reused?

Rohan138 replied:
PettingZoo supports agents being done at different timesteps and, for example, removed from the environment; so `possible_agents` is a fixed tuple of all possible agent IDs, but if an agent is done before the other agents/environment is, we remove that agent ID from the `agents` list.

duenez replied:
OK, makes sense!

duenez commented on examples/pettingzoo/sb3_train.py:
```diff
@@ -0,0 +1,166 @@
+import supersuit as ss
+import torch
+import torch.nn.functional as F
+from stable_baselines3 import PPO
```
As above, import modules, nos classes.

duenez reviewed (changes requested):
Thank you very much for this contribution. Apologies for the long turnaround time. We were getting guidance as to how thorough we should be making the reviews of substantial code contributions to the repo. We are able to make minor edits to contributions from GitHub to make sure our code is compliant with internal formatting and style guidelines. Since this was a more substantial contribution, we felt it would be too difficult to fix all comments on our side. The decision, thus, was to provide a full code review, with the standards for internal code. We appreciate your patience while we did this.

Let me just say that we do appreciate this contribution, and feel it would add value to the project! Thanks!

I've added extensive comments, and some suggestions for refactoring the code, and possibly add tests to some shared pieces. I understand some of that code was just lifted from the RLLib example, but this would make the structure much better. Please let me know if the change is not trivial, of if there are unexpected blockers and we can re-evaluate the level of change we would need for this code to go in.

Hope that makes sense, and if you have any questions or comment, please let me know and would be more than happy to chat.

Thanks again!

Rohan138 commented:
Hi @duenez, thanks for the detailed review!

I've fixed the module imports, and added some responses about the PettingZoo API. I hope this answers your questions; feel free to let me know if there's anything else you'd like me to change or answer.

Thank you for your time!

duenez reviewed (changes requested):
Thank you for the changes. Just a minor comment to clean up the petting zoo environment module, otherwise this looks fine to me :+1:

duenez reviewed (dismissed):
Thanks again for the contribution and patience!

Rohan138 commented:
Sure, thank you for your time!

duenez commented:
This is now merged into the main branch of the repo.
